snuba/utils/streams/batching.py

```diff
@@ -61,6 +62,150 @@ class Offsets:
+                "partition": str(message.partition.index),
+            },
+        )
+
+        # Create the batch only after the first message is seen.
+        if self.__batch is None:
+            self.__batch = Batch()
+
+        result = self.__worker.process_message(message)
+
+        # XXX: ``None`` is indistinguishable from a potentially valid return
+        # value of ``TResult``!
+        if result is not None:
+            self.__batch.results.append(result)
+
+        duration = (time.time() - start) * 1000
+        self.__batch.messages_processed_count += 1
+        self.__batch.processing_time_ms += duration
+        self.__metrics.timing("process_message", duration)
+
+        if message.partition in self.__batch.offsets:
+            self.__batch.offsets[message.partition].hi = message.get_next_offset()
+        else:
+            self.__batch.offsets[message.partition] = Offsets(
+                message.offset, message.get_next_offset()
+            )
+
+    def close(self) -> None:
+        """
+        Close the processor, discarding any messages (without committing
+        offsets) that were previously consumed and processed since the last
+        batch flush.
+        """
+        self.__closed = True
+
+    def __flush(self) -> None:
+        """
+        Flush the active batch and reset the batch state.
+        """
+        assert not self.__closed
+        assert self.__batch is not None, "cannot flush without active batch"
+
+        logger.info(
+            "Flushing %s items (from %r)",
+            len(self.__batch.results),
+            self.__batch.offsets,
+        )
+
+        self.__metrics.timing(
+            "process_message.normalized",
+            self.__batch.processing_time_ms / self.__batch.messages_processed_count,
+        )
+
+        batch_results_length = len(self.__batch.results)
+        if batch_results_length > 0:
+            logger.debug("Flushing batch via worker")
+            flush_start = time.time()
+            self.__worker.flush_batch(self.__batch.results)
+            flush_duration = (time.time() - flush_start) * 1000
+            logger.info("Worker flush took %dms", flush_duration)
+            self.__metrics.timing("batch.flush", flush_duration)
+            self.__metrics.timing(
+                "batch.flush.normalized", flush_duration / batch_results_length
+            )
+
+        logger.debug("Committing offsets for batch")
+        commit_start = time.time()
+        self.__consumer.stage_offsets(
+            {
+                partition: offsets.hi
+                for partition, offsets in self.__batch.offsets.items()
+            }
+        )
+        offsets = self.__consumer.commit_offsets()
+        logger.debug("Committed offsets: %s", offsets)
+        commit_duration = (time.time() - commit_start) * 1000
+        logger.debug("Offset commit took %dms", commit_duration)
+
+        self.__batch = None
+
+
 class BatchingConsumer(Generic[TPayload]):
```

Comment:
Taking suggestions for the future name of this class. Please don't let me name it `ConsumerManager`, but I have no other ideas what to call it yet. It will take the `Consumer[TPayload]` and `ProcessorFactory[TPayload]` and manage the processor lifecycle across assignments (basically what is happening here, but more abstract.)

Reply:
> I assume this ame should not hint at "batching" anymore, right? That responsibility is now entirely on the processor. Is this correct?

Yes, this is accurate. This is on track to be cleaned up in the follow up change, which renames this class to `StreamProcessor`. I didn't want to this `Consumer` since that makes it kind of confusing with the `KafkaConsumer` and generic abstract `streams.consumers.Consumer` classes. (I'm not really sure another variant of `Processor` is all that much better, but I think `MessageProcessor` might go away or be renamed by the time I am done with all of these changes.)

As far as the `Processor` itself (not concretely in this change but in the follow up), I think `ProcessingStrategy` might make sense (e.g. `BatchProcessingStrategy`) though the instance variable noun form would be kind of strange in context (`self.__processing_strategy.process(…)`) but not all that hard to live with.